stress: failed test in cockroach/kv/kv.test: TestSendRPCRetry

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/54be70507f68d5745312defbbf0ace8ada1a6660

Stress build found a failed test:

```
=== RUN   TestSendRPCRetry
I160819 04:05:03.689232 gossip/simulation/network.go:68  simulating gossip network with 1 nodes
W160819 04:05:03.689451 gossip/gossip.go:1039  not connected to cluster; use --join to specify a connected node
I160819 04:05:03.691097 stopper.go:408  quiesceing; tasks left:
9      gossip/infostore.go:290
I160819 04:05:03.691140 stopper.go:408  quiesceing; tasks left:
8      gossip/infostore.go:290
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x8ae24a]

goroutine 472 [running]:
panic(0x15893c0, 0xc4200100e0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Recover(0xc4202e0b40)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:176 +0x6e
panic(0x15893c0, 0xc4200100e0)
	/usr/local/go/src/runtime/panic.go:458 +0x243
google.golang.org/grpc.NewClientStream(0x7eff807580f8, 0xc4201b2000, 0x2027d20, 0x0, 0x16e0355, 0x1f, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/google.golang.org/grpc/stream.go:113 +0x1ea
github.com/cockroachdb/cockroach/gossip.(*gossipClient).Gossip(0xc42016a010, 0x7eff807580f8, 0xc4201b2000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/gossip/gossip.pb.go:170 +0xb5
github.com/cockroachdb/cockroach/gossip.(*client).start.func1.2(0xc42021c000, 0x600c01)
	/go/src/github.com/cockroachdb/cockroach/gossip/client.go:99 +0x14f
github.com/rubyist/circuitbreaker.(*Breaker).Call(0xc42021c000, 0xc4201b22c0, 0x0, 0xc42078e0d0, 0xc4207da030)
	/go/src/github.com/rubyist/circuitbreaker/circuitbreaker.go:339 +0x6b
github.com/cockroachdb/cockroach/gossip.(*client).start.func1()
	/go/src/github.com/cockroachdb/cockroach/gossip/client.go:103 +0x277
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc4202e0b40, 0xc4209ba480)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x7d
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x66

goroutine 1 [chan receive]:
testing.(*T).Run(0xc4201c80c0, 0x16d01a8, 0x10, 0x1818ff8, 0xc4202d3c01)
	/usr/local/go/src/testing/testing.go:647 +0x316
testing.RunTests.func1(0xc4201c80c0)
	/usr/local/go/src/testing/testing.go:793 +0x6d
testing.tRunner(0xc4201c80c0, 0xc42004be00)
	/usr/local/go/src/testing/testing.go:610 +0x81
testing.RunTests(0x181b6a0, 0x2036320, 0x83, 0x83, 0x23ca528)
	/usr/local/go/src/testing/testing.go:799 +0x2f5
testing.(*M).Run(0xc42004bef8, 0x23ca100)
	/usr/local/go/src/testing/testing.go:743 +0x85
github.com/cockroachdb/cockroach/kv_test.TestMain(0xc42004bef8)
	/go/src/github.com/cockroachdb/cockroach/kv/main_test.go:53 +0x76
main.main()
	github.com/cockroachdb/cockroach/kv/_test/_testmain.go:318 +0xc6

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2086 +0x1

goroutine 5 [chan receive]:
github.com/cockroachdb/cockroach/util/log.(*loggingT).flushDaemon(0x237cbe0)
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:1016 +0x77
created by github.com/cockroachdb/cockroach/util/log.init.1
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:599 +0x95

goroutine 347 [runnable]:
fmt.(*pp).free(0xc420288600)
	/usr/local/go/src/fmt/print.go:141
fmt.Sprintf(0x16c3bc1, 0x5, 0xc420963a48, 0x2, 0x2, 0xc420148260, 0xade32ea4ab421401)
	/usr/local/go/src/fmt/print.go:198 +0xb8
github.com/cockroachdb/cockroach/util/stop.taskKey.String(0x1d26cee, 0x13, 0x122, 0xc420148310, 0xc42018ed80)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:95 +0xe8
github.com/cockroachdb/cockroach/util/stop.(*Stopper).runningTasksLocked(0xc4202e0b40, 0x2)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:328 +0x18d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Quiesce(0xc4202e0b40)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:408 +0x145
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Stop(0xc4202e0b40)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:352 +0x17f
github.com/cockroachdb/cockroach/gossip/simulation.(*Network).Stop(0xc4200f4580)
	/go/src/github.com/cockroachdb/cockroach/gossip/simulation/network.go:216 +0x8f
github.com/cockroachdb/cockroach/gossip/simulation.(*Network).Stop-fm()
	/go/src/github.com/cockroachdb/cockroach/kv/dist_sender_test.go:132 +0x2a
github.com/cockroachdb/cockroach/kv.TestSendRPCRetry(0xc4200a6540)
	/go/src/github.com/cockroachdb/cockroach/kv/dist_sender_test.go:894 +0x838
testing.tRunner(0xc4200a6540, 0x1818ff8)
	/usr/local/go/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:646 +0x2ec

goroutine 514 [runnable]:
sync.runtime_Semacquire(0xc4200a015c)
	/usr/local/go/src/runtime/sema.go:47 +0x30
sync.(*Mutex).Lock(0xc4200a0158)
	/usr/local/go/src/sync/mutex.go:85 +0xd0
github.com/cockroachdb/cockroach/gossip.(*infoStore).runCallbacks.func2()
	/go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:278 +0x3e
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask.func1(0xc4202e0b40, 0x1d26cee, 0x13, 0x122, 0xc420011ed0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:231 +0x95
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:232 +0xbc

goroutine 515 [runnable]:
net.uitoa(0x9627, 0x90f244, 0xc42078c188)
	/usr/local/go/src/net/parse.go:202
net.itoa(0x9627, 0x4, 0x4)
	/usr/local/go/src/net/parse.go:198 +0x94
net.(*TCPAddr).String(0xc420236990, 0xc42091e026, 0xa)
	/usr/local/go/src/net/tcpsock.go:33 +0x13e
net.(*OpError).Error(0xc420066f00, 0x1818ba8, 0xc420294540)
	/usr/local/go/src/net/net.go:443 +0xcf
fmt.(*pp).handleMethods(0xc420294540, 0xc400000076, 0x624801)
	/usr/local/go/src/fmt/print.go:584 +0x1e2
fmt.(*pp).printArg(0xc420294540, 0x15ab2a0, 0xc420066f00, 0x76)
	/usr/local/go/src/fmt/print.go:665 +0x17b
fmt.(*pp).doPrintf(0xc420294540, 0x16d933f, 0x19, 0xc42091e030, 0x1, 0x1)
	/usr/local/go/src/fmt/print.go:985 +0x123d
fmt.Sprintf(0x16d933f, 0x19, 0xc42091e030, 0x1, 0x1, 0x7eff8084af01, 0xc420930d90)
	/usr/local/go/src/fmt/print.go:196 +0x6a
golang.org/x/net/trace.(*eventLog).printf(0xc4201f0000, 0x601b00, 0x16d933f, 0x19, 0xc42091e030, 0x1, 0x1)
	/go/src/golang.org/x/net/trace/events.go:319 +0x9f
golang.org/x/net/trace.(*eventLog).Printf(0xc4201f0000, 0x16d933f, 0x19, 0xc42091e030, 0x1, 0x1)
	/go/src/golang.org/x/net/trace/events.go:311 +0x62
google.golang.org/grpc.(*Server).printf(0xc4201f6500, 0x16d933f, 0x19, 0xc42091e030, 0x1, 0x1)
	/go/src/google.golang.org/grpc/server.go:217 +0x7b
google.golang.org/grpc.(*Server).Serve(0xc4201f6500, 0x204cae0, 0xc42076e008, 0x0, 0x0)
	/go/src/google.golang.org/grpc/server.go:351 +0x245
github.com/cockroachdb/cockroach/gossip/simulation.(*Network).StartNode.func1()
	/go/src/github.com/cockroachdb/cockroach/gossip/simulation/network.go:135 +0x40
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc4202e0b40, 0xc420011ee0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x7d
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x66

goroutine 518 [semacquire]:
sync.runtime_Semacquire(0xc4200a015c)
	/usr/local/go/src/runtime/sema.go:47 +0x30
sync.(*Mutex).Lock(0xc4200a0158)
	/usr/local/go/src/sync/mutex.go:85 +0xd0
github.com/cockroachdb/cockroach/gossip.(*infoStore).runCallbacks.func2()
	/go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:278 +0x3e
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask.func1(0xc4202e0b40, 0x1d26cee, 0x13, 0x122, 0xc420011f50)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:231 +0x95
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:232 +0xbc

goroutine 349 [chan receive]:
github.com/cockroachdb/cockroach/gossip/simulation.(*Network).CreateNode.func1()
	/go/src/github.com/cockroachdb/cockroach/gossip/simulation/network.go:108 +0xcf
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc4202e0b40, 0xc4202369f0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x7d
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x66

goroutine 521 [semacquire]:
sync.runtime_Semacquire(0xc4200a015c)
	/usr/local/go/src/runtime/sema.go:47 +0x30
sync.(*Mutex).Lock(0xc4200a0158)
	/usr/local/go/src/sync/mutex.go:85 +0xd0
github.com/cockroachdb/cockroach/gossip.(*infoStore).runCallbacks.func2()
	/go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:278 +0x3e
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask.func1(0xc4202e0b40, 0x1d26cee, 0x13, 0x122, 0xc4202a0900)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:231 +0x95
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:232 +0xbc

goroutine 522 [semacquire]:
sync.runtime_Semacquire(0xc4200a015c)
	/usr/local/go/src/runtime/sema.go:47 +0x30
sync.(*Mutex).Lock(0xc4200a0158)
	/usr/local/go/src/sync/mutex.go:85 +0xd0
github.com/cockroachdb/cockroach/gossip.(*infoStore).runCallbacks.func2()
	/go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:278 +0x3e
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask.func1(0xc4202e0b40, 0x1d26cee, 0x13, 0x122, 0xc4202a0a20)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:231 +0x95
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:232 +0xbc

goroutine 519 [semacquire]:
sync.runtime_Semacquire(0xc4200a015c)
	/usr/local/go/src/runtime/sema.go:47 +0x30
sync.(*Mutex).Lock(0xc4200a0158)
	/usr/local/go/src/sync/mutex.go:85 +0xd0
github.com/cockroachdb/cockroach/gossip.(*infoStore).runCallbacks.func2()
	/go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:278 +0x3e
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask.func1(0xc4202e0b40, 0x1d26cee, 0x13, 0x122, 0xc4202a0430)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:231 +0x95
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:232 +0xbc

goroutine 520 [semacquire]:
sync.runtime_Semacquire(0xc4200a015c)
	/usr/local/go/src/runtime/sema.go:47 +0x30
sync.(*Mutex).Lock(0xc4200a0158)
	/usr/local/go/src/sync/mutex.go:85 +0xd0
github.com/cockroachdb/cockroach/gossip.(*infoStore).runCallbacks.func2()
	/go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:278 +0x3e
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask.func1(0xc4202e0b40, 0x1d26cee, 0x13, 0x122, 0xc4202a0840)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:231 +0x95
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:232 +0xbc

goroutine 516 [semacquire]:
sync.runtime_Semacquire(0xc4200a015c)
	/usr/local/go/src/runtime/sema.go:47 +0x30
sync.(*Mutex).Lock(0xc4200a0158)
	/usr/local/go/src/sync/mutex.go:85 +0xd0
github.com/cockroachdb/cockroach/gossip.(*infoStore).runCallbacks.func2()
	/go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:278 +0x3e
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask.func1(0xc4202e0b40, 0x1d26cee, 0x13, 0x122, 0xc420011f00)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:231 +0x95
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:232 +0xbc

goroutine 352 [runnable]:
github.com/cockroachdb/cockroach/gossip.(*Gossip).manage.func1()
	/go/src/github.com/cockroachdb/cockroach/gossip/gossip.go:917 +0x2fa
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc4202e0b40, 0xc420011ac0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x7d
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x66

goroutine 351 [select]:
github.com/cockroachdb/cockroach/gossip.(*Gossip).bootstrap.func1()
	/go/src/github.com/cockroachdb/cockroach/gossip/gossip.go:882 +0x2af
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc4202e0b40, 0xc420011a10)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x7d
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x66

goroutine 348 [runnable]:
sync.(*Mutex).Unlock(0xc4201c4360)
	/usr/local/go/src/sync/mutex.go:102
sync.(*Once).Do(0xc4201c4360, 0x18197d0)
	/usr/local/go/src/sync/once.go:46 +0x94
github.com/cockroachdb/cockroach/rpc.NewContext.func1()
	/go/src/github.com/cockroachdb/cockroach/rpc/context.go:127 +0x111
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc4202e0b40, 0xc42078c120)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x7d
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x66


ERROR: exit status 2
```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
0 runs so far, 0 failures, over 15s
0 runs so far, 0 failures, over 20s
0 runs so far, 0 failures, over 25s
5 runs so far, 0 failures, over 30s
16 runs so far, 0 failures, over 35s
16 runs so far, 0 failures, over 40s
16 runs so far, 0 failures, over 45s
16 runs so far, 0 failures, over 50s
16 runs so far, 0 failures, over 55s
31 runs so far, 0 failures, over 1m0s
32 runs so far, 0 failures, over 1m5s
32 runs so far, 0 failures, over 1m10s
32 runs so far, 0 failures, over 1m15s
32 runs so far, 0 failures, over 1m20s
33 runs so far, 0 failures, over 1m25s
48 runs so far, 0 failures, over 1m30s
48 runs so far, 0 failures, over 1m35s
48 runs so far, 0 failures, over 1m40s
48 runs so far, 0 failures, over 1m45s
48 runs so far, 0 failures, over 1m50s
55 runs so far, 0 failures, over 1m55s
64 runs so far, 0 failures, over 2m0s
64 runs so far, 0 failures, over 2m5s
64 runs so far, 0 failures, over 2m10s
64 runs so far, 0 failures, over 2m15s
65 runs so far, 0 failures, over 2m20s
80 runs so far, 0 failures, over 2m25s
80 runs so far, 0 failures, over 2m30s
80 runs so far, 0 failures, over 2m35s
80 runs so far, 0 failures, over 2m40s
81 runs so far, 0 failures, over 2m45s
84 runs so far, 0 failures, over 2m50s
95 runs completed, 1 failures, over 2m52s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/rpc/context.go
+++ b/rpc/context.go
@@ -124,7 +124,14 @@ func NewContext(baseCtx *base.Context, hlcClock *hlc.Clock, stopper *stop.Stoppe
 		ctx.cancel()
 		ctx.conns.Lock()
 		for key, meta := range ctx.conns.cache {
-			meta.Do(func() {}) // Make sure initialization is not in-progress.
+			meta.Do(func() {
+				// Make sure initialization is not in progress when we're removing the
+				// conn. We need to set the error in case we win the race against the
+				// real initialization code.
+				if meta.err == nil {
+					meta.err = &roachpb.NodeUnavailableError{}
+				}
+			})
 			ctx.removeConnLocked(key, meta)
 		}
 		ctx.conns.Unlock()

